discovery/source/DiscoveryClient.cpp

```diff
@@ -0,0 +1,165 @@
+            tlsConnectionOptions.SetServerName(serverName);
+
+            Crt::Http::HttpClientConnectionManagerOptions connectionManagerOptions;
+            connectionManagerOptions.socketOptions = clientConfig.socketOptions;
+            connectionManagerOptions.bootstrap = clientConfig.bootstrap;
+            connectionManagerOptions.tlsConnectionOptions = &tlsConnectionOptions;
+            connectionManagerOptions.maxConnections = clientConfig.maxConnections;
+            connectionManagerOptions.initialWindowSize = SIZE_MAX;
+            connectionManagerOptions.hostName = serverName;
+            connectionManagerOptions.port = port;
+
+            m_connectionManager = Crt::Http::HttpClientConnectionManager::NewClientConnectionManager(
+                connectionManagerOptions, clientConfig.allocator);
+        }
+
+        struct ClientCallbackContext
+        {
+            Crt::StringStream ss;
+            int responseCode;
+        };
+
+        bool DiscoveryClient::Discover(
+            const Crt::String &thingName,
+            const OnDiscoverResponse &onDiscoverResponse) noexcept
+        {
+
+            auto *callbackContext = Crt::New<ClientCallbackContext>(m_allocator);
+            if (!callbackContext)
+            {
+                return false;
+            }
+
+            callbackContext->responseCode = 0;
+
+            bool res = m_connectionManager->AcquireConnection(
+                [this, callbackContext, thingName, onDiscoverResponse](
+                    std::shared_ptr<Crt::Http::HttpClientConnection> connection, int errorCode) {
+                    if (!errorCode)
+                    {
+                        Crt::StringStream ss;
+                        ss << "/greengrass/discover/thing/" << thingName;
+                        Crt::String uriStr = ss.str();
+                        Crt::Http::HttpRequestOptions requestOptions;
+                        requestOptions.uri = Crt::ByteCursorFromCString(uriStr.c_str());
+                        requestOptions.method = Crt::ByteCursorFromCString("GET");
+
+                        Crt::Http::HttpHeader hostNameHeader;
+                        hostNameHeader.name = Crt::ByteCursorFromCString("host");
+                        hostNameHeader.value = Crt::ByteCursorFromCString(m_hostName.c_str());
+
+                        requestOptions.headerArray = &hostNameHeader;
+                        requestOptions.headerArrayLength = 1;
+
+                        requestOptions.onStreamOutgoingBody = nullptr;
+                        requestOptions.onIncomingHeaders =
+                            [](Crt::Http::HttpStream &, const Crt::Http::HttpHeader *, std::size_t) {};
+                        requestOptions.onIncomingHeadersBlockDone =
+                            [callbackContext](Crt::Http::HttpStream &stream, bool) {
+                                callbackContext->responseCode = stream.GetResponseStatusCode();
```

Comment:
[optional] this callback could be null and you could just query the response-code in the onStreamComplete() callback

Reply:
I wanted the ability to branch on it before writing to the stream. When we start generating this stuff I'll prolly revisit it.